Lib/threading.py

```diff
@@ -368,16 +368,14 @@ def notify(self, n=1):
         """
         if not self._is_owned():
             raise RuntimeError("cannot notify on un-acquired lock")
-        all_waiters = self._waiters
-        waiters_to_notify = _deque(_islice(all_waiters, n))
-        if not waiters_to_notify:
+        waiters = self._waiters
+
+        if not waiters:
             return
-        for waiter in waiters_to_notify:
+        num_to_notify = min(n, len(waiters))
+        for i in range(num_to_notify):
+            waiter = waiters.popleft()
```

Comment:
The try-except in the old code around `all_waiters.remove` concerns me slightly, since it implies that `self._waiters` can racily lose elements.

Maybe we should guard against waiters being empty `if not waiters: break`?

Reply:
I think waiters must be guarded by lock.

Reply:
I'm not sure we need a lock (beyond the GIL). Just need to make sure if enough waiters get removed elsewhere that there are no longer `num_to_notify` we don't crash.

Reply:
The previous version guarded it with a lock *and* had the try-except.  The reason I didn't use try-except was because having both seemed redundant to me.  The only way I can think of that `self._waiters` could racily lose elements is if a different function removed an element without the lock, which isn't done in any of the methods in `Condition`.

If preferable, I could add the try-except back.